src/util/prolog/prologSession.ts

```diff
@@ -1,12 +1,32 @@
+import fs from "fs";
+import path from "path";
 import { format_answer, type } from "tau-prolog/modules/core";
 import {PrologResult} from "./prologResult";
 
+// RegExp group one will be the file to be imported. Note: not global
+const fetchPrologImportsRegExp = /:-( |)use_module\(([a-zA-Z1-9]*)\)\./;
+
+export function importFile(relativePath: string): PrologSession {
+	const absolutePath = path.resolve(__dirname, relativePath);
+	const program = fs.readFileSync(absolutePath, "utf-8");
+	return new PrologSession(program, path.dirname(absolutePath));
+}
+
 export class PrologSession {
 
 		public readonly session: any;
 
-		constructor(program: string, limit?: number) {
+		constructor(program: string, defaultPath: string = __dirname, limit?: number) {
 				this.session = new type.Session(limit);
+
+				let matchGroups: string[] | null;
+				let importedCode: string;
+
+				while ((matchGroups = fetchPrologImportsRegExp.exec(program)) !== null) {
+					 importedCode = fs.readFileSync(defaultPath + path.sep + matchGroups[2] + ".pl", "utf-8");
+					 program = program.replace(fetchPrologImportsRegExp, importedCode);
+				}
```

Comment:
I understand you implement the importing by inlining. But if A imports B and C, and both B and C import D, D is included twice in your result. You should consider adding a `Set<string>` of filenames to your algorithm that you already imported.

Reply:
I guess the order in which stuff is eventually imported doesn't matter? The prolog-programs lines can be reordered however?

Reply:
You don't need to use `exec` and `replace` both, but you can do instead something like this:

```typescript
var program = program.replace(fetchPrologImportsRegExp, match => {
    return fs.readFileSync(defaultPath + path.sep + match[2] + ".pl", "utf-8");
});
```

Reply:
The order of the imports almost doesn't matter. There are examples where an import in a different order would break the code or change the order of the answers. If runtime can be improved by not strictly keeping the import order it would probably make sense to do so.

Reply:
If the order of imports *does* matter, you probably want to sort all the imports based on their partial order. The algorithm for that is simple: Determine the complete dependency graph of all imports (A imports B, B imports C, B imports C, ...), then start picking files (A, B, C ...) that have no dependencies in the graph (import other stuff). Then remove that node and all import-edges. Add the contents of the files to your result in the order you pick them from the graph. *Tada*  
This will solve the problems with duplicate imports and circular dependencies will be obvious to spot.